dask/bag/core.py

```diff
@@ -817,16 +817,34 @@ def topk(self, k, key=None, split_every=None):
         return self.reduction(func, compose(func, toolz.concat), out_type=Bag,
                               split_every=split_every, name='topk')
 
-    def distinct(self):
+    def distinct(self, key=None):
         """ Distinct elements of collection
 
         Unordered without repeats.
 
+        Parameters
+        ----------
+        key: {callable,str}
+            Defines uniqueness of items in bag by calling `key` on each item.
+            If a string is passed `key` is considered to be `lambda x: x[key]`.
+
+        Examples
+        --------
         >>> b = from_sequence(['Alice', 'Bob', 'Alice'])
         >>> sorted(b.distinct())
         ['Alice', 'Bob']
+        >>> b = from_sequence([{'name': 'Alice'}, {'name': 'Bob'}, {'name': 'Alice'}])
+        >>> b.distinct(key=lambda x: x['name']).compute()
+        [{'name': 'Alice'}, {'name': 'Bob'}]
+        >>> b.distinct(key='name').compute()
+        [{'name': 'Alice'}, {'name': 'Bob'}]
         """
-        return self.reduction(set, merge_distinct, out_type=Bag,
+        key_func = key if callable(key) or key is None else lambda x: x[key]
+        perpartition = lambda seq: list(toolz.unique(seq, key=key_func))
```

Comment:
We'd like to avoid the dynamic creation of functions here.  They are somewhat costly to serialize.  I recommend making another top level function like `merge_distinct`